### DISCUSSION TITLE
Multi-user support for shiori

### DISCUSSION THREAD

**Discussion by hulb** [go-shiori/shiori#516] [Feature requests]
As for now(v1.5.3), every user in shiori sees the same content.I think it would be better to seperate user's content by providing a multi-user mechanism and every user would have his/her own account and data.

**Comment by fmartingr** [comment]
I actually thought this was already proposed! But yeah, on this one I agree. It should be done with care though, for users that already have "guests".

**Reply by distresscode** [reply]
I got on here to open a ticket about this. It's odd that the data is shared for all users. Can someone please explain how the "guest" role works? Thanks!

**Reply by fmartingr** [reply]
Hey @distresscode,

right now all bookmarks are shared with all users, the **owners** can add/edit bookmarks while **guests** can only see them (but not add new/modify).

This was created as a single-user software first, as a way to have a bookmark silo that can be shared with others.

Changing this wont be easy though.

**Comment by gemorgan** [comment]
I agree with having multi-user support. I actually just added my wife as a user thinking she could store her own bookmarks but realized it doesn't work that way yet.  Seems like it shouldn't be too hard to work in, adding a user_id column to bookmarks and settings. I'm not sure that how the ebooks and images are stored needs to be modified for this.
One additional feature to keep in mind when doing it is the ability to allow sharing of a bookmark with another user. I think a simple sharing table should handle that.

**Reply by Monirzadeh** [reply]
Current visitor user can access to which one?
Should user each time select who can accesses which?
visitor user can see who share the article in this system?
If user save same URL we should save them separately?(for example 2 user save same URL with 1 year later and thumbnail change how we can handle that not affect each other) If not how we should control accesses to the file on server like epub,etc.
If one of them update cache what should be happen for other one?
Should we keep duplicate content in database?
What do you thinking about them?

**Reply by fmartingr** [reply]
My thoughts on this (realted to this) is that with a multi-user workflow the bookmarks can be private (only accesible to the user who created it, default), internal (accessible to all users of the server), public (accessible to all users of the server and to non-logged in users). Something like that.

Regarding the database and internals, is not that simple and it would probably involve a multi-step release/migration, not only to reference users around but to properly remove the uniqueness of URLs, solve backward compatbile issues (for servers that have more than one admin user), etc. 

This is something that I want to do, but there are other things that I want to tackle first (API, Import/Export, Archival).

**Comment by ckuethe** [comment]
I like the Private/Internal/Public separation. Some possible user roles:

* Unauth: unauthed random internet user. Can only read Public bookmarks
* Guest: maybe rename it to "read-only", to indicate that they can not create new bookmarks but can read "Internal" bookmarks, and perhaps their old "Private" bookmarks if they had some but were demoted from user.
* User: can create Internal or Private bookmarks
* Admin: can create other users, can modify other users' permissions, can modify other users' bookmarks, can create Public bookmarks
